File: leetcode/medium/Kth_Smallest_Element_In_a_BST.py
# -*- coding: utf-8 -*-

"""
created by huash06 at 2015-07-07 22:34

Given a binary search tree, write a function kthSmallest to find the kth smallest element in it.

Note:
You may assume k is always valid, 1 ≤ k ≤ BST's total elements.

Follow up:
What if the BST is modified (insert/delete operations) often and you need to find the kth smallest frequently?
How would you optimize the kthSmallest routine?

Hint:

Try to utilize the property of a BST.
What if you could modify the BST node's structure?
The optimal runtime complexity is O(height of BST).
"""

__author__ = 'huash06'

import sys
import os
import itertools
import collections
import functools
import bisect
import datetime


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    # @param {TreeNode} root
    # @param {integer} k
    # @return {integer}
    def kthSmallest(self, root, k):

        return self.dfs(root, k)[0]

    def dfs(self, root, k):
        if not root:
            return 0, k
        v, lk = self.dfs(root.left, k)
        if lk == 0:
            return v, 0
        if lk == 1:
            return root.val, 0
        return self.dfs(root.right, lk - 1)


root = TreeNode(5)
root.left = TreeNode(3)
root.right = TreeNode(8)
root.right.left = TreeNode(7)
root.right.right = TreeNode(9)

s = Solution()
for i in range(1, 6):
    print(s.kthSmallest(root, i))

File: cpp/leetcodes/Kth_Smallest_Element_in_a_BST.cpp
/*
Given a binary search tree, write a function kthSmallest to find the kth smallest element in it.

Note:
You may assume k is always valid, 1 ? k ? BST's total elements.

Follow up:
What if the BST is modified (insert/delete operations) often and you need to find the kth smallest frequently? How would you optimize the kthSmallest routine?

Hint:

Try to utilize the property of a BST.
What if you could modify the BST node's structure?
The optimal runtime complexity is O(height of BST).
 */

#include <iostream>
#include <vector>
#include <bits/stl_algo.h>
#include <set>
#include <queue>
#include <stack>
using namespace std;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        //pre-order travel
        stack<TreeNode*> q;
        TreeNode* node = root;
        while (node != nullptr){
            q.push(node);
            node = node->left;
        }
        while (!q.empty()){
            node = q.top();
            k--;
            if (k == 0){
                return node->val;
            }
            q.pop();
            if(node->right != nullptr){
                node = node->right;
                while (node != nullptr){
                    q.push(node);
                    node = node->left;
                }
            }
        }
        return -1;
    }
};

int main(){
    TreeNode* root = new TreeNode(6);
    root->left = new TreeNode(4);
    root->left->left = new TreeNode(3);
    root->left->right = new TreeNode(5);
    root->right = new TreeNode(8);
    root->right->left = new TreeNode(7);
    root->right->right = new TreeNode(9);
    Solution s;
    for(int i=1; i <= 7; i++){
        cout << s.kthSmallest(root, i) << endl;
    }
}